src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -0,0 +1,411 @@
+import de.tum.in.www1.artemis.domain.LearningObject;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.competency.Competency;
+import de.tum.in.www1.artemis.domain.competency.CompetencyRelation;
+import de.tum.in.www1.artemis.domain.competency.LearningPath;
+import de.tum.in.www1.artemis.repository.*;
+import de.tum.in.www1.artemis.web.rest.dto.PageableSearchDTO;
+import de.tum.in.www1.artemis.web.rest.dto.SearchResultPageDTO;
+import de.tum.in.www1.artemis.web.rest.dto.learningpath.NgxLearningPathDTO;
+import de.tum.in.www1.artemis.web.rest.util.PageUtil;
+
+@Service
+public class LearningPathService {
+
+    private final Logger log = LoggerFactory.getLogger(LearningPathService.class);
+
+    private final UserRepository userRepository;
+
+    private final LearningPathRepository learningPathRepository;
+
+    private final CompetencyProgressRepository competencyProgressRepository;
+
+    private final CourseRepository courseRepository;
+
+    private final CompetencyRelationRepository competencyRelationRepository;
+
+    private final LectureUnitRepository lectureUnitRepository;
+
+    private final ExerciseRepository exerciseRepository;
+
+    public LearningPathService(UserRepository userRepository, LearningPathRepository learningPathRepository, CompetencyProgressRepository competencyProgressRepository,
+            CourseRepository courseRepository, CompetencyRelationRepository competencyRelationRepository, LectureUnitRepository lectureUnitRepository,
+            ExerciseRepository exerciseRepository) {
+        this.userRepository = userRepository;
+        this.learningPathRepository = learningPathRepository;
+        this.competencyProgressRepository = competencyProgressRepository;
+        this.courseRepository = courseRepository;
+        this.competencyRelationRepository = competencyRelationRepository;
+        this.lectureUnitRepository = lectureUnitRepository;
+        this.exerciseRepository = exerciseRepository;
+    }
+
+    /**
+     * Generate learning paths for all students enrolled in the course
+     *
+     * @param course course the learning paths are created for
+     */
+    public void generateLearningPaths(@NotNull Course course) {
+        var students = userRepository.getStudents(course);
+        students.forEach((student) -> generateLearningPathForUser(course, student));
+        log.debug("Successfully created learning paths for all {} students in course (id={})", students.size(), course.getId());
+    }
+
+    /**
+     * Generate learning path for the user in the course
+     *
+     * @param course course that defines the learning path
+     * @param user   student for which the learning path is generated
+     */
+    public void generateLearningPathForUser(@NotNull Course course, @NotNull User user) {
+        var existingLearningPath = learningPathRepository.findByCourseIdAndUserId(course.getId(), user.getId());
+        if (existingLearningPath.isEmpty()) {
```

Comment:
Personal preference, especially if the method gets longer in future PRs:
```suggestion
        if (existingLearningPath.isPresent()) {
            return;
        }
```